esp-hal/src/dma/mod.rs

```diff
@@ -607,32 +575,69 @@ macro_rules! dma_descriptors_chunk_size {
 #[macro_export]
 macro_rules! dma_circular_descriptors_chunk_size {
     ($rx_size:expr, $tx_size:expr, $chunk_size:expr) => {{
-        // these will check for size at compile time
-        const _: () = ::core::assert!($chunk_size <= 4095, "chunk size must be <= 4095");
-        const _: () = ::core::assert!($chunk_size > 0, "chunk size must be > 0");
+        $crate::dma_descriptors_impl!($rx_size, $tx_size, $chunk_size, is_circular = true)
+    }};
 
-        const rx_descriptor_len: usize = if $rx_size > $chunk_size * 2 {
-            ($rx_size + $chunk_size - 1) / $chunk_size
-        } else {
-            3
-        };
+    ($size:expr, $chunk_size:expr) => {
+        $crate::dma_circular_descriptors_chunk_size!($size, $size, $chunk_size)
+    };
+}
+
+#[doc(hidden)]
+#[macro_export]
+macro_rules! dma_buffers_impl {
+    ($rx_size:expr, $tx_size:expr, $chunk_size:expr, is_circular = $circular:tt) => {{
+        let rx = $crate::dma_buffers_impl!($rx_size, $chunk_size, is_circular = $circular);
+        let tx = $crate::dma_buffers_impl!($tx_size, $chunk_size, is_circular = $circular);
+        (rx.0, rx.1, tx.0, tx.1)
+    }};
 
-        const tx_descriptor_len: usize = if $tx_size > $chunk_size * 2 {
-            ($tx_size + $chunk_size - 1) / $chunk_size
+    ($size:expr, $chunk_size:expr, is_circular = $circular:tt) => {{
+        $crate::declare_aligned_dma_buffer!(BUFFER, $size);
+
+        unsafe {
+            (
+                $crate::as_mut_byte_array!(BUFFER, $size),
+                $crate::dma_descriptors_impl!($size, $chunk_size, is_circular = $circular),
+            )
+        }
+    }};
+}
+
+#[doc(hidden)]
+#[macro_export]
+macro_rules! dma_descriptors_impl {
+    (@count $size:expr, $chunk_size:expr, is_circular = true) => {
+        if $size > $chunk_size * 2 {
+            ($size as usize).div_ceil($chunk_size)
         } else {
             3
```

Comment:
Maybe a smidge clearer?
```suggestion
        ($size as usize).div_ceil($chunk_size).max(3)
```

Reply:
Unfortunately I keep reading "but max 3" in these cases so maybe not 🫠